File: management-domain/Finance-department/Backend/Nodes/finance-automation-service/payment-automation-service/src/shared/base/base.entity.ts
import { v4 as uuidv4 } from 'uuid';

export abstract class BaseEntity {
  id: string;
  createdAt: Date;
  updatedAt: Date;
  tenantId: string;
  version: number;

  protected constructor(tenantId?: string) {
    this.id = uuidv4();
    this.createdAt = new Date();
    this.updatedAt = new Date();
    this.tenantId = tenantId || 'default';
    this.version = 1;
  }

  updateTimestamp(): void {
    this.updatedAt = new Date();
    this.version++;
  }

  equals(other: BaseEntity): boolean {
    if (other === this) return true;
    if (other === null || other === undefined) return false;
    return this.id === other.id;
  }
}

File: management-domain/Customer-support/Customer-support/Backend/Java/country-support-dashboard-service/src/main/java/com/gogidix/customersupport/countrysupportdashboard/domain/model/BaseEntity.java
package com.gogidix.customersupport.countrysupportdashboard.domain.model;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Field;
import java.time.Instant;
import java.util.Objects;

public abstract class BaseEntity {
    @Id protected String id;
    @Indexed @Field("tenant_id") protected String tenantId;
    @Field("created_at") protected Instant createdAt;
    @Field("updated_at") protected Instant updatedAt;

    protected BaseEntity() {
        this.createdAt = Instant.now();
        this.updatedAt = Instant.now();
    }

    protected BaseEntity(String tenantId) {
        this.id = java.util.UUID.randomUUID().toString();
        this.tenantId = Objects.requireNonNull(tenantId, "tenantId is required");
        this.createdAt = Instant.now();
        this.updatedAt = Instant.now();
    }

    public String getId() { return id; }
    public String getTenantId() { return tenantId; }
    public Instant getCreatedAt() { return createdAt; }
    public Instant getUpdatedAt() { return updatedAt; }
    public void setId(String id) { this.id = id; }
    public void setTenantId(String tenantId) { this.tenantId = tenantId; }
    public void setCreatedAt(Instant createdAt) { this.createdAt = createdAt; }
    public void setUpdatedAt(Instant updatedAt) { this.updatedAt = updatedAt; }

    public void updateTimestamp() { this.updatedAt = Instant.now(); }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        BaseEntity that = (BaseEntity) o;
        return Objects.equals(id, that.id);
    }

    @Override
    public int hashCode() { return Objects.hash(id); }
}
